File: Breadth-First Search/String/LET_no127_Word Ladder.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        word_memo = set(wordList)
        if endWord not in word_memo:
            return 0
        
        queue = collections.deque([beginWord])
        visited = set([beginWord])
        times = 1
        
        while queue:
            for _ in range(len(queue)):
                cur = queue.popleft()
                if cur == endWord:
                    return times
                for nxt in self.get_next_words(cur, word_memo):
                    if nxt not in visited:
                        visited.add(nxt)
                        queue.append(nxt)
            times += 1
        
        return 0
    
    def get_next_words(self, w, word_memo):
        res = []
        for i in range(len(w)):
            for d in range(26):
                c = chr(ord('a') + d)
                if c == w[i]:
                    continue
                nxt = w[:i] + c + w[i+1:]
                if nxt in word_memo:
                    res.append(nxt)
        
        return res

File: Breadth-First Search/String/LET_no127_Word Ladder.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> memo(wordList.begin(), wordList.end());
        if (!memo.count(endWord))
            return 0;

        unordered_set<string> visited({beginWord});
        queue<string> q;
        q.push(beginWord);
        int times = 1;

        while (!q.empty()) {
            int n = q.size();
            while (n-- > 0) {
                string cur = q.front();
                q.pop();

                if (cur == endWord)
                    return times;
                
                for (int i = 0; i < cur.size(); i++) {
                    char c = cur[i];
                    
                    for (int d = 0; d < 26; d++) {
                        char c2 = 'a' + d;
                        if (c2 == cur[i])
                            continue;
                        
                        cur[i] = c2;
                        if (memo.count(cur) && !visited.count(cur)) {
                            visited.insert(cur);
                            q.push(cur);
                        }
                    }

                    cur[i] = c;
                }
            }

            times++;
        }

        return 0;
    }
};